### PR TITLE
Navigation: implement a follow location + destination mode

### PR DESCRIPTION
**Pull request by nirvn**
This PR imrpoves QField's navigation functionality by implementing a follow location + destination mode to the map canvas. When enabled, the map canvas will automatically re-center to include both the current location point as well as destination. As the location gets closer to the destination, the map canvas zooms in.

UI/UX:

https://user-images.githubusercontent.com/1728657/159675089-7012c82a-6cd9-453b-af04-e413adee0e2c.mp4

Basically, if the user clicks on the GPS/location button to enable auto-follow, the navigation button (when a destination is declared) can be used as a toggle to enable / disable this new auto follow mode.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 6cba3472b806a2854a0f9885a88439f5a9d8b30d:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-e137a542dbad6ae5c67bf6da3124942e2aa6f789-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-e137a542dbad6ae5c67bf6da3124942e2aa6f789-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-e137a542dbad6ae5c67bf6da3124942e2aa6f789-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-e137a542dbad6ae5c67bf6da3124942e2aa6f789-x86-android.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 5e9f2fb92c53f3c5e679b47c86406941b84865ae:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-05dadcc47017e534f19f1f3521fb85299ad2907f-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-05dadcc47017e534f19f1f3521fb85299ad2907f-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-05dadcc47017e534f19f1f3521fb85299ad2907f-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-05dadcc47017e534f19f1f3521fb85299ad2907f-x86-android.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 53dd6575b21222e7d198858490b5cbd2049fad73:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-d971a1e4f08d59d5f435a184ca671b27d5964496-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-d971a1e4f08d59d5f435a184ca671b27d5964496-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-d971a1e4f08d59d5f435a184ca671b27d5964496-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-d971a1e4f08d59d5f435a184ca671b27d5964496-x86-android.apk)

**Comment by nirvn** [comment]
How it looks in motion:

https://user-images.githubusercontent.com/1728657/159703310-c15f26f3-ab68-45a1-8015-f7849fd73bee.mp4

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 53dd6575b21222e7d198858490b5cbd2049fad73:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-56a5c20a4eaab3b952aa0f44a1d0564a5bd51963-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-56a5c20a4eaab3b952aa0f44a1d0564a5bd51963-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-56a5c20a4eaab3b952aa0f44a1d0564a5bd51963-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-56a5c20a4eaab3b952aa0f44a1d0564a5bd51963-x86-android.apk)

File: src/core/qgsquick/qgsquickmapsettings.h [hunk 1]
```diff
@@ -160,12 +162,17 @@ class QFIELD_CORE_EXPORT QgsQuickMapSettings : public QObject
     //! Move current map extent to have center point defined by \a layer. Optionally only pan to the layer if \a shouldZoom is false.
     Q_INVOKABLE void setCenterToLayer( QgsMapLayer *layer, bool shouldZoom = true );
 
+    Q_INVOKABLE void setExtentFromPoints( const QVariantList &points );
```
Comment (suricactus) [inline src/core/qgsquick/qgsquickmapsettings.h:165]: sorry for being that guy, but this is the only method in the neighbouring lines without doc.

File: src/qml/qgismobileapp.qml [hunk 2]
```diff
@@ -1159,6 +1169,33 @@ ApplicationWindow {
...
+      function followLocation(forceRecenter) {
+        var screenLocation = mapCanvas.mapSettings.coordinateToScreen(positionSource.projectedPosition);
+        if (navigation.isActive && navigationButton.followIncludeDestination) {
+          if (mapCanvas.mapSettings.scale > 10) {
+            var screenDestination = mapCanvas.mapSettings.coordinateToScreen(navigation.destination);
+            if (forceRecenter || screenDestination.x < 40 || screenDestination.x > (mainWindow.width - 40) ||
+                screenDestination.y < 40 || screenDestination.y > (mainWindow.height - 40) ||
+                screenLocation.x < 40 || screenLocation > (mainWindow.width - 40) ||
+                screenLocation.y < 40 || screenLocation.y > (mainWindow.height - 40) ||
+                (Math.abs(screenDestination.x - screenLocation.x) < mainWindow.width / 3 && Math.abs(screenDestination.y - screenLocation.y) < mainWindow.height / 3)) {
```
Comment (suricactus) [inline src/qml/qgismobileapp.qml:1181]:
```suggestion
            if (
                forceRecenter
                || screenDestination.x < 40
                || screenDestination.x > (mainWindow.width - 40)
                || screenDestination.y < 40
                || screenDestination.y > (mainWindow.height - 40)
                || screenLocation.x < 40
                || screenLocation > (mainWindow.width - 40)
                || screenLocation.y < 40
                || screenLocation.y > (mainWindow.height - 40)
                || (
                    Math.abs(screenDestination.x - screenLocation.x) < mainWindow.width / 3
                    && Math.abs(screenDestination.y - screenLocation.y) < mainWindow.height / 3
                )
            ) {
```

better.

File: src/qml/qgismobileapp.qml [hunk 3]
```diff
@@ -1159,6 +1169,33 @@ ApplicationWindow {
         gpsMenu.popup(locationToolbar.x + locationToolbar.width - gpsMenu.width, locationToolbar.y + locationToolbar.height - gpsMenu.height)
       }
 
+      function followLocation(forceRecenter) {
```
Comment (suricactus) [inline src/qml/qgismobileapp.qml:1172]: I see some magic number 10 and 40. I think they can be converted to more descriptive constants. What I understood is:

```
FOLLOW_LOCATION_MIN_SCALE = 10
FOLLOW_LOCATION_MIN_SCREEN_MARGIN = 40
```

File: src/qml/qgismobileapp.qml [hunk 4]
```diff
@@ -1159,6 +1169,33 @@ ApplicationWindow {
...
+              var points = [positionSource.projectedPosition, navigation.destination];
+              mapCanvas.mapSettings.setExtentFromPoints(points)
+            }
+          }
+        } else {
+          var screenFraction = settings.value( "/QField/Positioning/FollowScreenFraction", 5 );
+          var threshold = Math.min( mainWindow.width, mainWindow.height ) / screenFraction;
+          if ( forceRecenter || screenLocation.x < threshold || screenLocation.x > mainWindow.width - threshold ||
+               screenLocation.y < threshold || screenLocation.y > mainWindow.height - threshold )
+          {
```
Comment (suricactus) [inline src/qml/qgismobileapp.qml:1192]:
```suggestion
          if (
            forceRecenter
            || screenLocation.x < threshold
            || screenLocation.x > mainWindow.width - threshold
            || screenLocation.y < threshold
            || screenLocation.y > mainWindow.height - threshold
          )
          {
```

I am now on smaller screens, it's huge pain to read and understand long lines.

File: src/qml/qgismobileapp.qml [hunk 5]
```diff
@@ -1007,17 +999,36 @@ ApplicationWindow {
       round: true
       anchors.right: parent.right
 
-      iconSource: Theme.getThemeIcon( "ic_navigation_flag_purple_24dp" )
-      bgcolor: Theme.darkGray
+      iconSource: positionSource.active && gpsButton.followActive && followIncludeDestination ? Theme.getThemeIcon( "ic_navigation_flag_white_24dp" ) : Theme.getThemeIcon( "ic_navigation_flag_purple_24dp" )
+      bgcolor: positionSource.active && gpsButton.followActive && followIncludeDestination ? Theme.navigationColor : Theme.darkGray
```
Comment (suricactus) [inline src/qml/qgismobileapp.qml:1003]:
```suggestion
      property bool isFollowLocationActive: positionSource.active && gpsButton.followActive && followIncludeDestination
      iconSource: isFollowLocationActive
        ? Theme.getThemeIcon( "ic_navigation_flag_white_24dp" )
        : Theme.getThemeIcon( "ic_navigation_flag_purple_24dp" )
      bgcolor: isFollowLocationActive
        ? Theme.navigationColor
        : Theme.darkGray
```

**Review by suricactus — APPROVED** [review]
Looks good, nice addtion. I put some comments on readability of very longs lines. Next month I will be extra sensitive about this, because I have a smaller screen.

**Comment by nirvn** [comment]
@suricactus , thanks for the review, comments addressed.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 37fb479b761b145fdc1e8fb5051f25c8cb01087e:
  - [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-458c66e9e9be897d692b586da796a2dbccf98adb-arm64-android.apk)

Other architectures: [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-458c66e9e9be897d692b586da796a2dbccf98adb-arm-android.apk), [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-458c66e9e9be897d692b586da796a2dbccf98adb-x64-android.apk), [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-2636-458c66e9e9be897d692b586da796a2dbccf98adb-x86-android.apk)
